File: gen.rb
#!/usr/bin/env ruby

text = File.read 'resume.md'
css  = File.read 'style.css'

output = text.scan(/(.+)\n(---+|===+)\n((?:(?:\t|\*).+\n?|\n)+)/).map { |title, line, contents|

	# process links
	contents.gsub!(/\[(.+?)\]\((.+?)\)/) { "<a href='#{$2}'>#{$1}</a>" }

	html = ''

	# add section headers (on side or <h1>)
	html     = "<div class=first>#{title}</div>"                      if line.match /---+/
	contents = "<div class=second><h1>#{title}</h1></div>" + contents if line.match(/===+/)

	# convert section content into seconds (with floating thirds)
	html + contents.gsub(/\* (.+)\n?((?:\t.+\n?)*)/) {

		main, sub = $1, $2

		sub.gsub!(/\t- (.+)/m) { $1.strip.gsub("\n", '<br />') }

		main.sub! /{(.+?)}/, ''
		lang = $1
		main.sub! /<([ \d\-Present,]+?)>$/, ''
		date = $1

		'<div class=second><div class=header>%s%s</div>%s</div>%s' % [
			main,
			("<span class=lang>#{lang}</span>" if lang),
			("<blockquote>#{sub}</blockquote>" if not sub.empty?),
			("<div class=third>#{date}</div>"  if date)
		]

	}

}

File.write 'docs/index.html', "<style>#{css.gsub /\s+/, ' '}</style><body>#{output.join '<div class=filler></div>'}</body>"

File: gen.js
#!/usr/bin/env -S deno run --allow-read --allow-write

const slurp = f => Deno.readFile(f).then(b => new TextDecoder().decode(b))

Function.prototype.after = function(f) {
	return (...args) => this(f(...args))
}

const opt = ([a, b], thing) => thing ? `${a}${thing}${b}` : ''

const contentf = (_, content, langm, datem) =>
	`<div class=content>${content}</div>
	${opt`<div class=lang>${langm}</div>`}
	${opt`<div class=date>${datem}</div>`}`

const process = text => text
	.replace(/(\w.+)\n===+/g, '<h1>$1</h1>')
	.replace(/(\w.+)\n---+/g, (_, one) => `<h2>${one}</h2>`)
	.replace(/\[(.+?)\]\((.+?)\)/g, "<a href='$2'>$1</a>")
	.replace(/^\* (.+?) ?(?:\{(.+)\})? ?(?:\<([Present\d- ]+)\>)?$/mg, contentf) // epic
	.replace(/^\t- (.+)/mg, '<blockquote class=detail>$1</blockquote>')

const text = await slurp('resume.md')
const output = text.split(/\n\n\n+/).map(process).join('<div class=spacer></div>')
const html = `<meta name=viewport content='width=device-width, initial-scale=1'><link href=style.css rel=stylesheet>` + output

await Deno.writeFile
	( 'docs/index.html'
	, new TextEncoder().encode(html)
	)
